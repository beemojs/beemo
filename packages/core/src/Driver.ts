import execa from 'execa';
import mergeWith from 'lodash/mergeWith';
import { Blueprint, isObject, optimal, Path, Predicates, predicates } from '@boost/common';
import { ConcurrentEvent, Event } from '@boost/event';
import { Plugin } from '@boost/plugin';
import {
  STRATEGY_COPY,
  STRATEGY_CREATE,
  STRATEGY_NATIVE,
  STRATEGY_NONE,
  STRATEGY_REFERENCE,
} from './constants';
import ConfigContext from './contexts/ConfigContext';
import DriverContext from './contexts/DriverContext';
import isClassInstance from './helpers/isClassInstance';
import Tool from './Tool';
import {
  Argv,
  BeemoTool,
  Driverable,
  DriverCommandOptions,
  DriverMetadata,
  DriverOptions,
  DriverOutput,
  DriverStrategy,
  Execution,
} from './types';

export default abstract class Driver<
    Config extends object = {},
    Options extends DriverOptions = DriverOptions
  >
  extends Plugin<BeemoTool, Options>
  implements Driverable {
  command: DriverCommandOptions = {};

  // Set after instantiation
  config!: Config;

  // Set after instantiation
  metadata!: DriverMetadata;

  // Set within a life-cycle
  tool!: Tool;

  output: DriverOutput = {
    stderr: '',
    stdout: '',
  };

  readonly onLoadProviderConfig = new Event<[ConfigContext, Path, Config]>('load-provider-config');

  readonly onLoadConsumerConfig = new Event<[ConfigContext, Config]>('load-consumer-config');

  readonly onMergeConfig = new Event<[ConfigContext, Config]>('merge-config');

  readonly onCreateConfigFile = new Event<[ConfigContext, Path, Config]>('create-config-file');

  readonly onCopyConfigFile = new Event<[ConfigContext, Path, Config]>('copy-config-file');

  readonly onReferenceConfigFile = new Event<[ConfigContext, Path, Config]>(
    'reference-config-file',
  );

  readonly onDeleteConfigFile = new Event<[ConfigContext, Path]>('delete-config-file');

  readonly onBeforeExecute = new ConcurrentEvent<[DriverContext, Argv]>('before-execute');

  readonly onAfterExecute = new ConcurrentEvent<[DriverContext, unknown]>('after-execute');

  readonly onFailedExecute = new ConcurrentEvent<[DriverContext, Error]>('failed-execute');

  static validate(driver: Driver) {
    const name = (isClassInstance(driver) && driver.constructor.name) || 'Driver';

    if (!isObject(driver.options)) {
      throw new Error(`\`${name}\` requires an options object.`);
    }
  }

  blueprint({ array, object, string }: Predicates): Blueprint<DriverOptions> {
    return {
      args: array(string()),
      dependencies: array(string()),
      env: object(string()),
      strategy: string(STRATEGY_NATIVE).oneOf<DriverStrategy>([
        STRATEGY_NATIVE,
        STRATEGY_CREATE,
        STRATEGY_REFERENCE,
        STRATEGY_COPY,
        STRATEGY_NONE,
      ]),
    };
  }

  bootstrap() {}

  startup(tool: Tool) {
    this.tool = tool;
    this.bootstrap();
  }

  /**
   * Special case for merging arrays.
   */
  doMerge(prevValue: unknown, nextValue: unknown): unknown {
    if (Array.isArray(prevValue) && Array.isArray(nextValue)) {
      return Array.from(new Set(prevValue.concat(nextValue)));
    }

    return undefined;
  }

  /**
   * Extract the error message when the driver fails to execute.
   */
  extractErrorMessage(error: { message: string }): string {
    return error.message.split('\n', 1)[0] || '';
  }

  /**
   * Format the configuration file before it's written.
   */
  formatConfig(data: Config): string {
    const content = JSON.stringify(data, null, 2);

    if (this.metadata.configName.endsWith('.js')) {
      return `module.exports = ${content};`;
    }

    return content;
  }

  /**
   * Return the module name without the Beemo namespace.
   */
  getName(): string {
    return this.name.split('-').pop()!;
  }

  /**
   * Return a list of user defined arguments.
   */
  getArgs(): Argv {
    return Array.isArray(this.options.args) ? this.options.args : [];
  }

  /**
   * Return a list of dependent drivers.
   */
  getDependencies(): string[] {
    const dependencies = Array.isArray(this.options.dependencies) ? this.options.dependencies : [];

    return [
      // Always required; configured by the driver
      ...this.metadata.dependencies,
      // Custom; configured by the consumer
      ...dependencies,
    ];
  }

  /**
   * Return a list of supported CLI options.
   */
  getSupportedOptions(): string[] {
    return [];
  }

  /**
   * Extract the current version of the installed driver via its binary.
   */
  getVersion(): string {
    const { bin, versionOption } = this.metadata;
    const version = (execa.sync(bin, [versionOption])?.stdout || '').trim();
    const match = version.match(/(\d+)\.(\d+)\.(\d+)/u);

    return match ? match[0] : '0.0.0';
  }

  /**
   * Merge multiple configuration objects.
   */
  mergeConfig(prev: Config, next: Config): Config {
    return mergeWith(prev, next, this.doMerge);
  }

  /**
   * Handle command failures according to this driver.
   */
  processFailure(error: Execution) {
    const { stderr = '', stdout = '' } = error;
    const out = (stderr || stdout).trim();

    if (out) {
      this.setOutput('stderr', out);
    }
  }

  /**
   * Handle successful commands according to this driver.
   */
  processSuccess(response: Execution) {
    const { stderr = '', stdout = '' } = response;

    this.setOutput('stderr', stderr.trim());
    this.setOutput('stdout', stdout.trim());
  }

  /**
   * Setup additional command options.
   */
  setCommandOptions(options: DriverCommandOptions): this {
    const blueprint: Blueprint<DriverCommandOptions> = {};
    const { shape, string } = predicates;

    Object.keys(options).forEach((key) => {
      blueprint[key] = shape({
        description: string().notEmpty().required(),
        type: string().oneOf<'boolean' | 'number' | 'string'>(['string', 'number', 'boolean']),
      });
    });

    this.command = optimal(options, blueprint, {
      name: this.constructor.name,
      unknown: true,
    });

    return this;
  }

  /**
   * Set metadata about the binary/executable in which this driver wraps.
   */
  setMetadata(metadata: Partial<DriverMetadata>): this {
    const { array, bool, string } = predicates;

    this.metadata = optimal(
      metadata,
      {
        bin: string()
          .match(/^[a-z]{1}[a-zA-Z0-9-]+$/u)
          .required(),
        configName: string().required(),
        configOption: string('--config'),
        configStrategy: string(STRATEGY_CREATE).oneOf([
          STRATEGY_CREATE,
          STRATEGY_REFERENCE,
          STRATEGY_COPY,
        ]),
        dependencies: array(string()),
        description: string(),
        filterOptions: bool(true),
        helpOption: string('--help'),
        title: string().required(),
        useConfigOption: bool(),
        versionOption: string('--version'),
        watchOptions: array(string()),
        workspaceStrategy: string(STRATEGY_REFERENCE).oneOf([STRATEGY_REFERENCE, STRATEGY_COPY]),
      },
      {
        name: this.constructor.name,
      },
    );

    return this;
  }

  /**
   * Store the raw output of the driver's execution.
   */
  setOutput(type: keyof DriverOutput, value: string): this {
    this.output[type] = value.trim();

    return this;
  }
}
