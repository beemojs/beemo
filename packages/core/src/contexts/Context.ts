import camelCase from 'lodash/camelCase';
import trim from 'lodash/trim';
import { PrimitiveType } from '@boost/args';
import { FilePath,Path } from '@boost/common';
import { Context as BaseContext } from '@boost/pipeline';
import { Arguments, Argv } from '../types';

export interface ConfigPath {
  driver: string;
  path: Path;
}

export default class Context<
  O extends object = {},
  P extends PrimitiveType[] = PrimitiveType[]
> extends BaseContext {
  // Parsed command line arguments as an object
  args: Arguments<O, P>;

  // Raw command line arguments as a list of strings
  argv: Argv;

  // List of driver configs currently registered
  configPaths: ConfigPath[] = [];

  // Current working directory
  cwd!: Path;

  // Absolute path to the configuration module
  configModuleRoot!: Path;

  // Absolute path to the folder containing `package.json` (Yarn workspaces) or `lerna.json`
  workspaceRoot!: Path;

  // List of paths (with trailing glob star) for each defined workspace
  workspaces: FilePath[] = [];

  constructor(args: Arguments<O, P>, argv: Argv = []) {
    super();

    this.args = args;
    this.argv = argv;

    // TODO MIGRATE
    // @ts-expect-error
    this.addArg = this.addParam.bind(this);
    // @ts-expect-error
    this.addArgs = this.addParams.bind(this);
  }

  /**
   * Add a config path for the defined driver.
   */
  addConfigPath(driverName: string, path: Path): this {
    this.configPaths.push({
      driver: driverName,
      path,
    });

    return this;
  }

  /**
   * Add an option argument to both the args object and argv list.
   * If the option is not supported, it will be added to the unknown options object.
   */
  addOption(arg: string, defaultValue: PrimitiveType = true, useEquals: boolean = false): this {
    let option = arg;
    let value = defaultValue;

    if (option.includes('=')) {
      [option, value] = option.split('=');
      useEquals = true;
    }

    let name = trim(option, '-');

    if (name.startsWith('no-')) {
      name = name.slice(3);
      value = false;
    }

    const optName = camelCase(name);

    if (optName in this.args.options) {
      // @ts-expect-error Allow this
      this.args.options[optName] = value;
    } else {
      this.args.unknown[optName] = String(value);
    }

    if (typeof value === 'boolean' || !value) {
      this.argv.push(option);
    } else if (useEquals) {
      this.argv.push(`${option}=${value}`);
    } else {
      this.argv.push(option, String(value));
    }

    return this;
  }

  /**
   * Add multiple boolean option arguments.
   */
  addOptions(args: string[]): this {
    args.forEach((arg) => {
      this.addOption(arg);
    });

    return this;
  }

  /**
   * Add a parameter to the argv list.
   */
  addParam(arg: string): this {
    this.args.params.push(arg);
    this.argv.push(arg);

    return this;
  }

  /**
   * Add multiple parameters.
   */
  addParams(args: string[]): this {
    args.forEach((arg) => {
      this.addParam(arg);
    });

    return this;
  }

  /**
   * Find a configuration path by file name.
   */
  findConfigByName(name: string): ConfigPath | undefined {
    return this.configPaths.find(
      (config) => String(config.path).endsWith(name) || config.driver === name,
    );
  }

  /**
   * Return a configured option value by name, or a fallback value if not found.
   */
  getOption<K extends keyof O>(name: K, fallback?: O[K]): O[K] | null {
    return this.args.options[name] || fallback || null;
  }

  /**
   * Return either a configured option value, or an unknown option value,
   * or null if not found.
   */
  getRiskyOption(name: string): PrimitiveType | null {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (this.getOption(name as keyof O) ?? this.args.unknown[name] ?? null) as any;
  }
}
