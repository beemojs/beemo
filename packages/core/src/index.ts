/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Blueprint, PackageStructure, Path, PortablePath, Predicates } from '@boost/common';

export * from './constants';
export * from './contexts/ConfigContext';
export * from './contexts/Context';
export * from './contexts/DriverContext';
export * from './contexts/ScaffoldContext';
export * from './contexts/ScriptContext';
export * from './Driver';
export * from './Script';
export * from './Tool';
export * from './types';

export { Path };
export type { Blueprint, PackageStructure, PortablePath, Predicates };
