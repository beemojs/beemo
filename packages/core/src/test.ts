import execa from 'execa';
import { ArgList, Arguments, Argv } from '@boost/args';
import { Path } from '@boost/common';
import { mockDebugger } from '@boost/debug/test';
import {
	ConfigContext,
	ConfigFile,
	Context,
	Driver,
	DriverContext,
	DriverContextOptions,
	DriverMetadata,
	ScaffoldContext,
	ScaffoldContextOptions,
	Script,
	ScriptContext,
	ScriptContextOptions,
	Tool,
} from '.';

export { mockDebugger };

const TEST_ROOT = new Path(process.env.BEEMO_TEST_ROOT ?? process.cwd());

export class TestDriver<O extends object = {}> extends Driver<O> {
	override readonly name = 'test-driver';
}

export class TestScript<O extends object = {}> extends Script<O> {
	override readonly name = 'test-script';

	execute(): any {
		return Promise.resolve();
	}
}

export function mockConsole<K extends keyof Console>(name: K): jest.SpyInstance {
	return jest.spyOn(console, name as 'log').mockImplementation(() => {});
}

export function mockToolConfig(): ConfigFile {
	return {
		configure: {
			cleanup: false,
			parallel: true,
		},
		debug: false,
		drivers: [],
		execute: {
			concurrency: 1,
			graph: true,
			output: '',
		},
		module: '@local',
		scripts: [],
		settings: {},
	};
}

export function mockTool(argv: Argv = []): Tool {
	const tool = new Tool({
		argv,
		cwd: TEST_ROOT,
	});

	// @ts-expect-error Allow readonly
	tool.debug = mockDebugger();

	tool.config = mockToolConfig();

	return tool;
}

export function mockDriver<C extends object = {}>(
	name: string,
	tool: Tool | null = null,
	metadata: Partial<DriverMetadata> = {},
): Driver<C> {
	const driver = new TestDriver<C>();

	// @ts-expect-error For testing purposes
	driver.name = name;
	driver.tool = tool ?? mockTool();

	driver.setMetadata({
		bin: name.toLowerCase(),
		configName: `${name}.json`,
		title: name,
		...metadata,
	});

	return driver;
}

export function mockScript(name: string, tool: Tool | null = null): Script<{}> {
	const script = new TestScript<{}>();

	// @ts-expect-error For testing purposes
	script.name = name;
	script.tool = tool ?? mockTool();

	return script;
}

export function stubArgs<T extends object>(options: T, params: ArgList = []): Arguments<T> {
	return {
		command: [],
		errors: [],
		options,
		params,
		rest: [],
		unknown: {},
	};
}

export function stubConfigArgs(): Arguments<{}> {
	return stubArgs({});
}

export function stubDriverArgs(
	fields?: Partial<DriverContextOptions>,
): Arguments<DriverContextOptions> {
	return stubArgs({
		concurrency: 1,
		graph: false,
		workspaces: '',
		...fields,
	});
}

export function stubScaffoldArgs(
	fields?: Partial<ScaffoldContextOptions>,
): Arguments<ScaffoldContextOptions> {
	return stubArgs({
		dry: false,
		...fields,
	});
}

export function stubScriptArgs(
	fields?: Partial<ScriptContextOptions>,
): Arguments<ScriptContextOptions> {
	return stubArgs({
		concurrency: 1,
		graph: false,
		workspaces: '',
		...fields,
	});
}

export function applyContext<T extends Context>(context: T): T {
	context.args = stubArgs({ a: true, foo: 'bar' }, ['baz']);
	context.argv = ['-a', '--foo', 'bar', 'baz'];
	context.cwd = TEST_ROOT;
	context.configModuleRoot = TEST_ROOT;
	context.workspaceRoot = TEST_ROOT;
	context.workspaces = [];

	return context;
}

export function stubConfigContext(): ConfigContext {
	return applyContext(new ConfigContext(stubArgs({})));
}

export function stubDriverContext(driver?: Driver): DriverContext {
	return applyContext(new DriverContext(stubDriverArgs(), driver ?? new TestDriver()));
}

export function stubScriptContext(script?: Script): ScriptContext {
	const context = applyContext(new ScriptContext(stubScriptArgs(), 'script'));

	if (script) {
		context.setScript(script);
	}

	return context;
}

export function stubScaffoldContext(
	generator: string = 'generator',
	action: string = 'action',
	name: string = '',
): ScaffoldContext {
	return applyContext(new ScaffoldContext(stubScaffoldArgs(), generator, action, name));
}

export function prependRoot(part: string): Path {
	return TEST_ROOT.append(part);
}

export function getRoot(): Path {
	return TEST_ROOT;
}

export function stubExecResult(fields?: Partial<execa.ExecaReturnValue>): execa.ExecaReturnValue {
	return {
		all: '',
		command: '',
		escapedCommand: '',
		exitCode: 0,
		failed: false,
		isCanceled: false,
		killed: false,
		signal: undefined,
		stderr: '',
		stdout: '',
		timedOut: false,
		...fields,
	};
}
