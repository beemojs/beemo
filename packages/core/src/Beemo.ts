import fs from 'fs-extra';
import path from 'path';
import { Event } from '@boost/event';
import { bool, number, string, shape } from 'optimal';
import CleanupConfigsRoutine from './routines/CleanupConfigsRoutine';
import ResolveConfigsRoutine from './routines/ResolveConfigsRoutine';
import RunDriverRoutine from './routines/RunDriverRoutine';
import Driver from './Driver';
import Context from './contexts/Context';
import DriverContext from './contexts/DriverContext';
import { Argv, Execution, BeemoConfig, BeemoPluginRegistry, DriverOptions } from './types';

export function configBlueprint() {
  return {
    configure: shape({
      cleanup: bool(false),
      parallel: bool(true),
    }),
    execute: shape({
      concurrency: number(),
      graph: bool(true),
    }),
    module: process.env.BEEMO_CONFIG_MODULE
      ? string(process.env.BEEMO_CONFIG_MODULE)
      : string().required(),
  };
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default class Beemo<T = any> extends Tool<BeemoPluginRegistry, BeemoConfig<T>> {
  onRunDriver = new Event<[DriverContext, Driver<object, DriverOptions>]>('run-driver');

  constructor(argv: Argv, binName?: string, testingOnly: boolean = false) {
    super(
      {
        appName: 'beemo',
        appPath: path.join(__dirname, '..'),
        configBlueprint: configBlueprint(),
        configName: binName || 'beemo',
        scoped: true,
      },
      argv,
    );

    // Abort early for testing purposes
    if (testingOnly) {
      return;
    }

    this.initialize();

    // Set footer after messages have been loaded
    const footer = this.msg('app:poweredBy', { version });

    this.options.footer = `\n${this.isCI() ? '' : 'ðŸ¤–  '}${footer}`;
  }

  /**
   * Execute all routines for the chosen driver.
   */
  async runDriver(
    args: DriverContext['args'],
    driverName: string,
    parallelArgv: Argv[] = [],
  ): Promise<Execution[]> {
    const driver = this.getPlugin('driver', driverName);
    const context = this.prepareContext(new DriverContext(args, driver, parallelArgv));
    const version = driver.getVersion();

    this.onRunDriver.emit([context, driver], driverName);

    this.debug('Running with %s v%s driver', driverName, version);

    const pipeline = this.startPipeline(context)
      .pipe(new ResolveConfigsRoutine('config', this.msg('app:configGenerate')))
      .pipe(
        new RunDriverRoutine(
          'driver',
          this.msg('app:driverRun', {
            name: driver.metadata.title,
            version,
          }),
        ),
      );

    // Only add cleanup routine if we need it
    if (this.config.configure.cleanup) {
      pipeline.pipe(new CleanupConfigsRoutine('cleanup', this.msg('app:cleanup')));
    }

    return pipeline.run(driverName);
  }

  /**
   * Delete config files if a process fails.
   */
  private handleCleanupOnFailure(code: number, context: Context) {
    if (code === 0) {
      return;
    }

    // Must not be async!
    if (Array.isArray(context.configPaths)) {
      context.configPaths.forEach((config) => {
        fs.removeSync(config.path.path());
      });
    }
  }
}
