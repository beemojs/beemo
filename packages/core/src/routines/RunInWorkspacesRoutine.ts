import Graph from '@beemo/dependency-graph';
import { Blueprint, PackageStructure,Predicates, WorkspacePackage } from '@boost/common';
import { PooledPipeline,Routine } from '@boost/pipeline';
import { stripAnsi, style } from '@boost/terminal';
import Context from '../contexts/Context';
import isPatternMatch from '../helpers/isPatternMatch';
import { ExecutionError, RoutineOptions } from '../types';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type AnyRoutine = Routine<any, any>;

export interface RunInWorkspacesContextArgs {
  concurrency: number;
  graph: boolean;
  workspaces: string;
}

const MAX_ERROR_LINES = 5;

export default abstract class RunInWorkspacesRoutine<
  Ctx extends Context<RunInWorkspacesContextArgs>
> extends Routine<unknown, unknown, RoutineOptions> {
  routines: AnyRoutine[] = [];

  workspacePackages: WorkspacePackage[] = [];

  blueprint({ instance }: Predicates): Blueprint<RoutineOptions> {
    return {
      // @ts-expect-error We cant import Tool because of cycles
      tool: instance().required().notNullable(),
    };
  }

  async execute(context: Ctx): Promise<unknown> {
    const { args, workspaces } = context;
    const { tool } = this.options;

    // Determine packages to run plugins in
    if (args.options.workspaces) {
      if (!workspaces || workspaces.length === 0) {
        throw new Error(tool.msg('errors:workspacesNotEnabled', { arg: args.options.workspaces }));
      }

      this.workspacePackages = tool.project.getWorkspacePackages();

      this.getFilteredWorkspacePackages(context).forEach((pkg) => {
        this.pipeRoutine(context, pkg.metadata.packageName, pkg.metadata.packagePath);
      });
    } else {
      this.pipeRoutine(context);
    }

    const value = await this.getInitialValue(context);
    const batches = this.orderByWorkspacePriorityGraph(context);
    const allErrors: Error[] = [];
    const allResults: unknown[] = [];
    const concurrency =
      context.args.options.concurrency || tool.config.execute.concurrency || undefined;

    // eslint-disable-next-line no-restricted-syntax
    for (const batch of batches) {
      const pipeline = batch.reduce(
        (pl, routine) => pl.add(routine),
        new PooledPipeline(context, value, { concurrency }),
      );

      // eslint-disable-next-line no-await-in-loop
      const { errors, results } = await pipeline.run();

      allResults.push(...results);

      if (errors.length > 0) {
        allErrors.push(...errors);
      }
    }

    if (allErrors.length > 0) {
      this.formatAndThrowErrors(allErrors);
    }

    // Not running in workspaces, so return value directly
    return context.args.options.workspaces ? allResults : allResults[0];
  }

  /**
   * When a list of errors are available, concatenate them and throw a new error.
   */
  formatAndThrowErrors(errors: Error[]) {
    let message = this.options.tool.msg('errors:executeFailed');

    (errors as ExecutionError[]).forEach((error) => {
      let content = stripAnsi(error.stderr || error.stdout || '')
        .trim()
        .split('\n');

      // istanbul ignore next
      if (content.length >= MAX_ERROR_LINES) {
        const count = content.length - MAX_ERROR_LINES;

        content = content.slice(0, MAX_ERROR_LINES);

        if (count > 0) {
          content.push(this.options.tool.msg('errors:executeFailedMoreLines', { count }));
        }
      }

      message += '\n\n';
      message += style.reset.yellow(error.message);

      if (content.length > 0) {
        message += '\n';
        message += style.reset.gray(content.join('\n'));
      }
    });

    message += '\n';

    const error = new Error(message);

    // Inherit stack for easier debugging.
    if (errors.length === 1) {
      error.stack = String(errors[0].stack).split('\n').slice(1).join('\n');
    }

    throw error;
  }

  /**
   * Return a list of workspaces optionally filtered.
   */
  getFilteredWorkspacePackages(context: Ctx): WorkspacePackage[] {
    return this.workspacePackages.filter((pkg) =>
      isPatternMatch(pkg.package.name, context.args.options.workspaces),
    );
  }

  /**
   * Return the initial value for the pipeline.
   */
  getInitialValue(context: Ctx): unknown {
    return null;
  }

  /**
   * Group routines in order of which they are dependend on.
   */
  orderByWorkspacePriorityGraph(context: Ctx): AnyRoutine[][] {
    const enabled = context.args.options.graph || this.options.tool.config.execute.graph || false;

    if (!enabled || !context.args.options.workspaces) {
      return [this.routines];
    }

    // Create lookups
    const packages: PackageStructure[] = []; // Without metadata
    const metadata: { [name: string]: WorkspacePackage['metadata'] } = {}; // By package name

    this.workspacePackages.forEach((wsp) => {
      packages.push(wsp.package);
      metadata[wsp.package.name] = wsp.metadata;
    });

    // Batch based on packages
    const batchList = new Graph(packages).resolveBatchList();
    const batches: AnyRoutine[][] = [];

    batchList.forEach((batch) => {
      const routines = batch
        .map((pkg) => this.routines.find((route) => route.key === metadata[pkg.name].packageName))
        .filter(Boolean) as AnyRoutine[];

      if (routines.length > 0) {
        batches.push(routines);
      }
    });

    return batches;
  }

  /**
   * Pipe a routine for the entire project or a workspace package at the defined path.
   */
  abstract pipeRoutine(context: Ctx, packageName?: string, packageRoot?: string): void;
}
