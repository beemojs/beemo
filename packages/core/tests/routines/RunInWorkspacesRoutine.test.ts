import { Predicates } from '@boost/common';
import { Routine } from '@boost/pipeline';
import Context from '../../src/contexts/Context';
import Driver from '../../src/Driver';
import RunInWorkspacesRoutine, {
  RunInWorkspacesContextArgs,
} from '../../src/routines/RunInWorkspacesRoutine';
import { mockDebugger, mockDriver, mockTool, stubDriverContext } from '../../src/test';
import Tool from '../../src/Tool';

type Ctx = Context<RunInWorkspacesContextArgs>;

class PipedRoutine extends Routine<unknown, unknown, { error?: Error; type?: string }> {
  blueprint({ instance, string }: Predicates) {
    return {
      error: instance(Error).nullable(),
      type: string(),
    };
  }

  execute() {
    const { error } = this.options;

    if (error) {
      if (this.options.type === 'stderr') {
        // @ts-expect-error
        error.stderr = 'Stderr message!';
      } else {
        // @ts-expect-error
        error.stdout = 'Stdout info...';
      }

      throw error;
    }

    return Promise.resolve(this.key);
  }
}

class ExecuteRoutine extends RunInWorkspacesRoutine<Ctx> {
  pipeRoutine(context: Ctx, packageName?: string, packageRoot?: string) {
    this.routines.push(new PipedRoutine(packageName || 'root', packageRoot || 'description'));
  }
}

describe('RunInWorkspacesRoutine', () => {
  let context: Ctx;
  let tool: Tool;
  let routine: RunInWorkspacesRoutine<Ctx>;
  let driver: Driver;
  let primary: PipedRoutine;
  let foo: PipedRoutine;
  let bar: PipedRoutine;
  let baz: PipedRoutine;
  let qux: PipedRoutine;

  beforeEach(() => {
    tool = mockTool();
    context = stubDriverContext(driver);
    driver = mockDriver('primary', tool);

    routine = new ExecuteRoutine('driver', 'Executing driver', { tool });
    // @ts-expect-error
    routine.debug = mockDebugger();

    // Setup packages
    primary = new PipedRoutine('primary', 'primary');
    foo = new PipedRoutine('foo', 'foo');
    bar = new PipedRoutine('bar', 'bar');
    baz = new PipedRoutine('baz', 'baz');
    qux = new PipedRoutine('qux', 'qux');

    routine.routines.push(primary);
    routine.routines.push(foo);
    routine.routines.push(bar);
    routine.routines.push(baz);
    routine.routines.push(qux);

    const packages = [
      {
        package: { name: '@scope/primary', version: '0.0.0' },
        metadata: tool.project.createWorkspaceMetadata('./packages/primary/package.json'),
      },
      {
        package: { name: '@scope/foo', version: '0.0.0' },
        metadata: tool.project.createWorkspaceMetadata('./packages/foo/package.json'),
      },
      {
        package: { name: '@scope/bar', version: '0.0.0' },
        metadata: tool.project.createWorkspaceMetadata('./packages/bar/package.json'),
      },
      {
        package: { name: '@scope/baz', version: '0.0.0' },
        metadata: tool.project.createWorkspaceMetadata('./packages/baz/package.json'),
      },
      {
        package: { name: '@scope/qux', version: '0.0.0' },
        metadata: tool.project.createWorkspaceMetadata('./packages/qux/package.json'),
      },
    ];

    routine.workspacePackages = packages;
    jest.spyOn(tool.project, 'getWorkspacePackages').mockImplementation(() => packages);
  });

  describe('execute()', () => {
    it('throws an error if any failures', async () => {
      routine.routines = [
        new PipedRoutine('out', 'Out', { error: new Error('Failed'), type: 'stdout' }),
        new PipedRoutine('err', 'Err', { error: new Error('Oops'), type: 'stderr' }),
      ];

      try {
        await routine.execute(context);
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
    });

    it('returns a result for a single routine', async () => {
      routine.routines = [foo];

      const response = await routine.execute(context);

      expect(response).toEqual('foo');
    });

    describe('workspaces', () => {
      beforeEach(() => {
        context.args.options.graph = true;
        context.args.options.workspaces = '*';
        context.workspaces = ['packages/*'];
      });

      it('returns an array of results for multiple routines', async () => {
        const response = await routine.execute(context);

        expect((response as string[]).sort()).toEqual(['bar', 'baz', 'foo', 'primary', 'qux']);
      });

      it('serializes priority routines before pooling other routines', async () => {
        const spy = jest.spyOn(routine, 'orderByWorkspacePriorityGraph');

        // primary -> foo
        routine.workspacePackages[0].package.peerDependencies = {
          '@scope/foo': '1.0.0',
        };

        await routine.execute(context);

        expect(spy).toHaveReturnedWith([[foo, bar, baz, qux], [primary]]);
      });
    });
  });

  describe('getFilteredWorkspaces()', () => {
    it('returns none for empty string', () => {
      context.args.options.workspaces = '';

      expect(routine.getFilteredWorkspacePackages(context)).toEqual([]);
    });

    it('returns all for wildcard `*`', () => {
      context.args.options.workspaces = '*';

      expect(routine.getFilteredWorkspacePackages(context)).toEqual([
        {
          package: { name: '@scope/primary', version: '0.0.0' },
          metadata: tool.project.createWorkspaceMetadata('./packages/primary/package.json'),
        },
        {
          package: { name: '@scope/foo', version: '0.0.0' },
          metadata: tool.project.createWorkspaceMetadata('./packages/foo/package.json'),
        },
        {
          package: { name: '@scope/bar', version: '0.0.0' },
          metadata: tool.project.createWorkspaceMetadata('./packages/bar/package.json'),
        },
        {
          package: { name: '@scope/baz', version: '0.0.0' },
          metadata: tool.project.createWorkspaceMetadata('./packages/baz/package.json'),
        },
        {
          package: { name: '@scope/qux', version: '0.0.0' },
          metadata: tool.project.createWorkspaceMetadata('./packages/qux/package.json'),
        },
      ]);
    });

    it('filters by package name', () => {
      context.args.options.workspaces = '@scope/(foo|bar)';

      expect(routine.getFilteredWorkspacePackages(context)).toEqual([
        {
          package: { name: '@scope/foo', version: '0.0.0' },
          metadata: tool.project.createWorkspaceMetadata('./packages/foo/package.json'),
        },
        {
          package: { name: '@scope/bar', version: '0.0.0' },
          metadata: tool.project.createWorkspaceMetadata('./packages/bar/package.json'),
        },
      ]);
    });

    it('filters by negation', () => {
      context.args.options.workspaces = '@scope/!(foo|baz)';

      expect(routine.getFilteredWorkspacePackages(context)).toEqual([
        {
          package: { name: '@scope/primary', version: '0.0.0' },
          metadata: tool.project.createWorkspaceMetadata('./packages/primary/package.json'),
        },
        {
          package: { name: '@scope/bar', version: '0.0.0' },
          metadata: tool.project.createWorkspaceMetadata('./packages/bar/package.json'),
        },
        {
          package: { name: '@scope/qux', version: '0.0.0' },
          metadata: tool.project.createWorkspaceMetadata('./packages/qux/package.json'),
        },
      ]);
    });
  });

  describe('orderByWorkspacePriorityGraph()', () => {
    beforeEach(() => {
      context.args.options.graph = true;
      context.args.options.workspaces = '*';
    });

    it('returns all in single batch if graph is false', () => {
      context.args.options.graph = false;
      tool.config.execute.graph = false;

      expect(routine.orderByWorkspacePriorityGraph(context)).toEqual([
        [primary, foo, bar, baz, qux],
      ]);
    });

    it('returns all in single batch if workspaces is empty', () => {
      context.args.options.workspaces = '';

      expect(routine.orderByWorkspacePriorityGraph(context)).toEqual([
        [primary, foo, bar, baz, qux],
      ]);
    });

    it('returns all in single batch if no dependents', () => {
      expect(routine.orderByWorkspacePriorityGraph(context)).toEqual([
        [bar, baz, foo, primary, qux],
      ]);
    });

    it('prioritizes based on peerDependencies', () => {
      // foo -> bar
      routine.workspacePackages[1].package.peerDependencies = {
        '@scope/bar': '1.0.0',
      };

      expect(routine.orderByWorkspacePriorityGraph(context)).toEqual([
        [bar, baz, primary, qux],
        [foo],
      ]);
    });

    it('prioritizes based on dependencies', () => {
      // foo -> bar
      routine.workspacePackages[1].package.dependencies = {
        '@scope/bar': '1.0.0',
      };

      expect(routine.orderByWorkspacePriorityGraph(context)).toEqual([
        [bar, baz, primary, qux],
        [foo],
      ]);
    });

    it('sorts priority based on dependency count', () => {
      // bar -> primary
      routine.workspacePackages[2].package.peerDependencies = {
        '@scope/primary': '2.0.0',
      };

      // foo -> bar
      routine.workspacePackages[1].package.dependencies = {
        '@scope/bar': '1.0.0',
      };

      // qux -> bar
      routine.workspacePackages[4].package.peerDependencies = {
        '@scope/bar': '1.0.0',
      };

      expect(routine.orderByWorkspacePriorityGraph(context)).toEqual([
        [primary, baz],
        [bar],
        [foo, qux],
      ]);
    });
  });
});
