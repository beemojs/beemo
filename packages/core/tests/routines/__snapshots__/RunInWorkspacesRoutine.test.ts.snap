// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RunInWorkspacesRoutine execute() throws an error if any failures 1`] = `
[Error: Failed to execute pipeline. The following errors have occurred:

[0m[33mFailed[39m[0m
[0m[90mStdout info...[39m[0m

[0m[33mOops[39m[0m
[0m[90mStderr message![39m[0m]
`;
